library/alloc/src/collections/vec_deque/drain.rs

```diff
@@ -108,57 +109,59 @@ impl<T, A: Allocator> Drop for Drain<'_, T, A> {
 
                 let source_deque = unsafe { self.0.deque.as_mut() };
 
-                let drain_start = source_deque.len();
                 let drain_len = self.0.drain_len;
-                let drain_end = drain_start + drain_len;
-
-                let orig_len = self.0.tail_len + drain_end;
+                let new_len = self.0.new_len;
 
                 if T::IS_ZST {
                     // no need to copy around any memory if T is a ZST
-                    source_deque.len = orig_len - drain_len;
+                    source_deque.len = new_len;
                     return;
                 }
 
-                let head_len = drain_start;
-                let tail_len = self.0.tail_len;
+                let head_len = source_deque.len();
+                let tail_len = new_len - head_len;
 
-                match (head_len, tail_len) {
-                    (0, 0) => {
-                        source_deque.head = 0;
-                        source_deque.len = 0;
-                    }
-                    (0, _) => {
-                        source_deque.head = source_deque.to_physical_idx(drain_len);
-                        source_deque.len = orig_len - drain_len;
-                    }
-                    (_, 0) => {
-                        source_deque.len = orig_len - drain_len;
+                // When draining at the front (`.drain(..n)`) or at the back (`.drain(n..)`),
+                // we don't need to copy any data.
+                // Outlining this function helps LLVM to eliminate the copies in these cases.
+                if head_len != 0 && tail_len != 0 {
+                    copy_data(source_deque, drain_len, head_len, tail_len);
+                }
+
+                if new_len == 0 {
+                    source_deque.head = 0;
+                } else if head_len < tail_len {
+                    source_deque.head = source_deque.to_physical_idx(drain_len);
+                }
+                source_deque.len = new_len;
+
+                #[cold]
+                fn copy_data<T, A: Allocator>(
```

Comment:
Maybe it could be called `shift_blocks` or `restore_wrapping` or something that describes its purpose rather than what it does internally?

It may copy the bits (in the `ptr::copy` sense, including potential overlap), but logically we don't want a copy, we want to shift the parts to restore deque invariants.